cluster.go

```diff
@@ -28,94 +40,119 @@ type ClusterClient struct {
-
-	return c.reset()
+	// TODO: close should make client unusable
+	return nil
 }
 
 // ------------------------------------------------------------------------
 
-// Finds the current master address for a given hash slot
-func (c *ClusterClient) getMasterAddrBySlot(hashSlot int) string {
-	if addrs := c.slots[hashSlot]; len(addrs) > 0 {
-		return addrs[0]
+// getClient returns a Client for a given address.
+func (c *ClusterClient) getClient(addr string) *Client {
+	c.clientsMx.RLock()
+	client, ok := c.clients[addr]
+	if ok {
+		c.clientsMx.RUnlock()
+		return client
 	}
-	return ""
-}
+	c.clientsMx.RUnlock()
 
-// Returns a node's client for a given address
-func (c *ClusterClient) getNodeClientByAddr(addr string) *Client {
-	c.connsMx.Lock()
-	client, ok := c.conns[addr]
+	c.clientsMx.Lock()
+	client, ok = c.clients[addr]
 	if !ok {
 		opt := c.opt.clientOptions()
 		opt.Addr = addr
 		client = NewTCPClient(opt)
-		c.conns[addr] = client
+		c.clients[addr] = client
 	}
-	c.connsMx.Unlock()
+	c.clientsMx.Unlock()
+
 	return client
 }
 
+// randomClient returns a Client for the first live node.
+func (c *ClusterClient) randomClient() *Client {
+	for i := 0; i < 10; i++ {
+		n := rand.Intn(len(c.addrs))
+		client := c.getClient(c.addrs[n])
+		if client.Ping().Err() == nil {
+			return client
+		}
+	}
+	return nil
+}
+
 // Process a command
 func (c *ClusterClient) process(cmd Cmder) {
+	var client *Client
 	var ask bool
 
 	c.reloadIfDue()
 
-	hashSlot := hashSlot(cmd.clusterKey())
+	slot := hashSlot(cmd.clusterKey())
 
 	c.slotsMx.RLock()
 	defer c.slotsMx.RUnlock()
 
-	tried := make(map[string]struct{}, len(c.addrs))
-	addr := c.getMasterAddrBySlot(hashSlot)
+	addrs := c.slots[slot]
+	if len(addrs) > 0 {
+		// First address is master.
+		client = c.getClient(addrs[0])
+	} else {
+		client = c.randomClient()
+	}
+
+	// Index in the addrs slice pointing to the next replica.
+	replicaIndex := 1
+
 	for attempt := 0; attempt <= c.opt.getMaxRedirects(); attempt++ {
-		tried[addr] = struct{}{}
+		if client == nil {
+			if cmd.Err() == nil {
+				err := fmt.Errorf("redis: no live nodes to serve this command")
+				cmd.setErr(err)
+			}
+			return
+		}
 
-		// Pick the connection, process request
-		conn := c.getNodeClientByAddr(addr)
 		if ask {
-			pipe := conn.Pipeline()
+			pipe := client.Pipeline()
 			pipe.Process(NewCmd("ASKING"))
 			pipe.Process(cmd)
 			_, _ = pipe.Exec()
 			ask = false
 		} else {
-			conn.Process(cmd)
+			client.Process(cmd)
 		}
 
 		// If there is no (real) error, we are done!
 		err := cmd.Err()
-		if err == nil || err == Nil {
+		if err == nil || err == Nil || err == TxFailedErr {
 			return
 		}
 
-		// On connection errors, pick a random, previosuly untried connection
-		// and request again.
-		if _, ok := err.(*net.OpError); ok || err == io.EOF {
-			if addr = c.findNextAddr(tried); addr == "" {
-				return
+		// On network errors try another node.
+		if isNetworkError(err) {
+			if replicaIndex < len(addrs) {
```

Comment:
Are replicas read-only? Otherwise I don't see reasons to not use them... AFAIK `randomClient` can pick replica too...